File: src/lib/bioUtils.ts
/* ============================================
   SYNTHETICA — Codon Table & Bio Utils
   Standard genetic code translation
   ============================================ */

// Standard Codon → Amino Acid mapping
export const CODON_TABLE: Record<string, string> = {
  // Phenylalanine
  'TTT': 'Phe', 'TTC': 'Phe',
  // Leucine
  'TTA': 'Leu', 'TTG': 'Leu', 'CTT': 'Leu', 'CTC': 'Leu', 'CTA': 'Leu', 'CTG': 'Leu',
  // Isoleucine
  'ATT': 'Ile', 'ATC': 'Ile', 'ATA': 'Ile',
  // Methionine (Start)
  'ATG': 'Met',
  // Valine
  'GTT': 'Val', 'GTC': 'Val', 'GTA': 'Val', 'GTG': 'Val',
  // Serine
  'TCT': 'Ser', 'TCC': 'Ser', 'TCA': 'Ser', 'TCG': 'Ser', 'AGT': 'Ser', 'AGC': 'Ser',
  // Proline
  'CCT': 'Pro', 'CCC': 'Pro', 'CCA': 'Pro', 'CCG': 'Pro',
  // Threonine
  'ACT': 'Thr', 'ACC': 'Thr', 'ACA': 'Thr', 'ACG': 'Thr',
  // Alanine
  'GCT': 'Ala', 'GCC': 'Ala', 'GCA': 'Ala', 'GCG': 'Ala',
  // Tyrosine
  'TAT': 'Tyr', 'TAC': 'Tyr',
  // Stop codons
  'TAA': 'Stop', 'TAG': 'Stop', 'TGA': 'Stop',
  // Histidine
  'CAT': 'His', 'CAC': 'His',
  // Glutamine
  'CAA': 'Gln', 'CAG': 'Gln',
  // Asparagine
  'AAT': 'Asn', 'AAC': 'Asn',
  // Lysine
  'AAA': 'Lys', 'AAG': 'Lys',
  // Aspartic acid
  'GAT': 'Asp', 'GAC': 'Asp',
  // Glutamic acid
  'GAA': 'Glu', 'GAG': 'Glu',
  // Cysteine
  'TGT': 'Cys', 'TGC': 'Cys',
  // Tryptophan
  'TGG': 'Trp',
  // Arginine
  'CGT': 'Arg', 'CGC': 'Arg', 'CGA': 'Arg', 'CGG': 'Arg', 'AGA': 'Arg', 'AGG': 'Arg',
  // Glycine
  'GGT': 'Gly', 'GGC': 'Gly', 'GGA': 'Gly', 'GGG': 'Gly',
};

// Amino acid properties (simplified for game mechanics)
export const AMINO_ACID_PROPERTIES: Record<string, {
  name: string;
  shortName: string;
  color: string;
  hydrophobic: boolean;
  heatStable: boolean;
  toxinResistant: boolean;
}> = {
  'Phe': { name: 'Phenylalanine', shortName: 'F', color: '#4a9eff', hydrophobic: true, heatStable: false, toxinResistant: false },
  'Leu': { name: 'Leucine', shortName: 'L', color: '#4a9eff', hydrophobic: true, heatStable: true, toxinResistant: false },
  'Ile': { name: 'Isoleucine', shortName: 'I', color: '#4a9eff', hydrophobic: true, heatStable: true, toxinResistant: false },
  'Met': { name: 'Methionine', shortName: 'M', color: '#ffaa00', hydrophobic: true, heatStable: false, toxinResistant: true },
  'Val': { name: 'Valine', shortName: 'V', color: '#4a9eff', hydrophobic: true, heatStable: true, toxinResistant: false },
  'Ser': { name: 'Serine', shortName: 'S', color: '#00ff88', hydrophobic: false, heatStable: false, toxinResistant: true },
  'Pro': { name: 'Proline', shortName: 'P', color: '#ff8844', hydrophobic: false, heatStable: true, toxinResistant: false },
  'Thr': { name: 'Threonine', shortName: 'T', color: '#00ff88', hydrophobic: false, heatStable: true, toxinResistant: true },
  'Ala': { name: 'Alanine', shortName: 'A', color: '#aaaaaa', hydrophobic: true, heatStable: true, toxinResistant: false },
  'Tyr': { name: 'Tyrosine', shortName: 'Y', color: '#00e5ff', hydrophobic: true, heatStable: false, toxinResistant: true },
  'His': { name: 'Histidine', shortName: 'H', color: '#00e5ff', hydrophobic: false, heatStable: false, toxinResistant: true },
  'Gln': { name: 'Glutamine', shortName: 'Q', color: '#00ff88', hydrophobic: false, heatStable: false, toxinResistant: true },
  'Asn': { name: 'Asparagine', shortName: 'N', color: '#00ff88', hydrophobic: false, heatStable: false, toxinResistant: false },
  'Lys': { name: 'Lysine', shortName: 'K', color: '#ff3366', hydrophobic: false, heatStable: false, toxinResistant: false },
  'Asp': { name: 'Aspartic Acid', shortName: 'D', color: '#ff3366', hydrophobic: false, heatStable: true, toxinResistant: false },
  'Glu': { name: 'Glutamic Acid', shortName: 'E', color: '#ff3366', hydrophobic: false, heatStable: true, toxinResistant: true },
  'Cys': { name: 'Cysteine', shortName: 'C', color: '#ffaa00', hydrophobic: true, heatStable: false, toxinResistant: true },
  'Trp': { name: 'Tryptophan', shortName: 'W', color: '#a855f7', hydrophobic: true, heatStable: false, toxinResistant: false },
  'Arg': { name: 'Arginine', shortName: 'R', color: '#ff3366', hydrophobic: false, heatStable: false, toxinResistant: true },
  'Gly': { name: 'Glycine', shortName: 'G', color: '#aaaaaa', hydrophobic: false, heatStable: true, toxinResistant: false },
  'Stop': { name: 'Stop', shortName: '*', color: '#666666', hydrophobic: false, heatStable: false, toxinResistant: false },
};

/**
 * Translate a DNA sequence into amino acids
 */
export function translateSequence(dna: string): string[] {
  const aminoAcids: string[] = [];
  const clean = dna.toUpperCase().replace(/[^ATCG]/g, '');
  
  for (let i = 0; i + 2 < clean.length; i += 3) {
    const codon = clean.substring(i, i + 3);
    const aa = CODON_TABLE[codon];
    if (aa) {
      if (aa === 'Stop') break;
      aminoAcids.push(aa);
    } else {
      aminoAcids.push('???');
    }
  }
  
  return aminoAcids;
}

/**
 * Get codons from a DNA sequence
 */
export function getCodons(dna: string): { codon: string; aminoAcid: string }[] {
  const codons: { codon: string; aminoAcid: string }[] = [];
  const clean = dna.toUpperCase().replace(/[^ATCG]/g, '');
  
  for (let i = 0; i + 2 < clean.length; i += 3) {
    const codon = clean.substring(i, i + 3);
    const aa = CODON_TABLE[codon] || '???';
    codons.push({ codon, aminoAcid: aa });
  }
  
  return codons;
}

/**
 * Calculate trait levels from amino acid sequence
 */
export function calculateTraits(aminoAcids: string[]): Record<string, number> {
  let heatScore = 0;
  let toxinScore = 0;
  let membraneScore = 0;
  let metabolicScore = 0;
  
  for (const aa of aminoAcids) {
    const props = AMINO_ACID_PROPERTIES[aa];
    if (!props) continue;
    
    if (props.heatStable) heatScore += 12;
    if (props.toxinResistant) toxinScore += 12;
    if (props.hydrophobic) membraneScore += 8;
    else membraneScore += 4;
    metabolicScore += 6;
  }
  
  // Normalize to 0-100
  const total = aminoAcids.length || 1;
  return {
    heat_resistance: Math.min(100, Math.round((heatScore / total) * 8)),
    toxin_resistance: Math.min(100, Math.round((toxinScore / total) * 8)),
    membrane_stability: Math.min(100, Math.round((membraneScore / total) * 10)),
    metabolic_efficiency: Math.min(100, Math.round((metabolicScore / total) * 12)),
  };
}

/**
 * Calculate folding stability from amino acid composition
 */
export function calculateFoldingStability(aminoAcids: string[]): number {
  if (aminoAcids.length === 0) return 0;
  
  let stability = 50;
  let hydrophobicRuns = 0;
  let currentRun = 0;
  
  for (const aa of aminoAcids) {
    const props = AMINO_ACID_PROPERTIES[aa];
    if (!props) continue;
    
    if (props.hydrophobic) {
      currentRun++;
      if (currentRun >= 3) hydrophobicRuns++;
    } else {
      currentRun = 0;
    }
    
    // Pro adds stability (rigid)
    if (aa === 'Pro') stability += 3;
    // Cys can form disulfide bonds
    if (aa === 'Cys') stability += 5;
    // Gly is too flexible
    if (aa === 'Gly') stability -= 2;
  }
  
  // Hydrophobic core is good
  stability += hydrophobicRuns * 5;
  
  return Math.max(0, Math.min(100, stability));
}

/**
 * Check if a nucleotide is valid
 */
export function isValidNucleotide(char: string): boolean {
  return ['A', 'T', 'C', 'G'].includes(char.toUpperCase());
}

/**
 * Validate a DNA sequence
 */
export function validateSequence(sequence: string): { valid: boolean; errors: string[] } {
  const errors: string[] = [];
  
  if (sequence.length === 0) {
    errors.push('Sequence is empty');
    return { valid: false, errors };
  }
  
  if (sequence.length < 3) {
    errors.push('Sequence must be at least 3 nucleotides (1 codon)');
  }
  
  for (let i = 0; i < sequence.length; i++) {
    if (!isValidNucleotide(sequence[i])) {
      errors.push(`Invalid nucleotide '${sequence[i]}' at position ${i + 1}`);
    }
  }
  
  if (sequence.length % 3 !== 0) {
    errors.push(`Sequence length (${sequence.length}) is not divisible by 3 — incomplete codon`);
  }
  
  return { valid: errors.length === 0, errors };
}

File: backend/bio_utils.py
# bio_utils.py
# Bioinformatics rules for sequence translation, folding stability, and trait calculations.

CODON_TABLE = {
    'TTT': 'Phe', 'TTC': 'Phe',
    'TTA': 'Leu', 'TTG': 'Leu', 'CTT': 'Leu', 'CTC': 'Leu', 'CTA': 'Leu', 'CTG': 'Leu',
    'ATT': 'Ile', 'ATC': 'Ile', 'ATA': 'Ile',
    'ATG': 'Met',
    'GTT': 'Val', 'GTC': 'Val', 'GTA': 'Val', 'GTG': 'Val',
    'TCT': 'Ser', 'TCC': 'Ser', 'TCA': 'Ser', 'TCG': 'Ser', 'AGT': 'Ser', 'AGC': 'Ser',
    'CCT': 'Pro', 'CCC': 'Pro', 'CCA': 'Pro', 'CCG': 'Pro',
    'ACT': 'Thr', 'ACC': 'Thr', 'ACA': 'Thr', 'ACG': 'Thr',
    'GCT': 'Ala', 'GCC': 'Ala', 'GCA': 'Ala', 'GCG': 'Ala',
    'TAT': 'Tyr', 'TAC': 'Tyr',
    'TAA': 'Stop', 'TAG': 'Stop', 'TGA': 'Stop',
    'CAT': 'His', 'CAC': 'His',
    'CAA': 'Gln', 'CAG': 'Gln',
    'AAT': 'Asn', 'AAC': 'Asn',
    'AAA': 'Lys', 'AAG': 'Lys',
    'GAT': 'Asp', 'GAC': 'Asp',
    'GAA': 'Glu', 'GAG': 'Glu',
    'TGT': 'Cys', 'TGC': 'Cys',
    'TGG': 'Trp',
    'CGT': 'Arg', 'CGC': 'Arg', 'CGA': 'Arg', 'CGG': 'Arg', 'AGA': 'Arg', 'AGG': 'Arg',
    'GGT': 'Gly', 'GGC': 'Gly', 'GGA': 'Gly', 'GGG': 'Gly',
}

AA_PROPERTIES = {
    'Phe': {'name': 'Phenylalanine', 'short': 'F', 'color': '#4a9eff', 'hydrophobic': True, 'heatStable': False, 'toxinResistant': False},
    'Leu': {'name': 'Leucine', 'short': 'L', 'color': '#4a9eff', 'hydrophobic': True, 'heatStable': True, 'toxinResistant': False},
    'Ile': {'name': 'Isoleucine', 'short': 'I', 'color': '#4a9eff', 'hydrophobic': True, 'heatStable': True, 'toxinResistant': False},
    'Met': {'name': 'Methionine', 'short': 'M', 'color': '#ffaa00', 'hydrophobic': True, 'heatStable': False, 'toxinResistant': True},
    'Val': {'name': 'Valine', 'short': 'V', 'color': '#4a9eff', 'hydrophobic': True, 'heatStable': True, 'toxinResistant': False},
    'Ser': {'name': 'Serine', 'short': 'S', 'color': '#00ff88', 'hydrophobic': False, 'heatStable': False, 'toxinResistant': True},
    'Pro': {'name': 'Proline', 'short': 'P', 'color': '#ff8844', 'hydrophobic': False, 'heatStable': True, 'toxinResistant': False},
    'Thr': {'name': 'Threonine', 'short': 'T', 'color': '#00ff88', 'hydrophobic': False, 'heatStable': True, 'toxinResistant': True},
    'Ala': {'name': 'Alanine', 'short': 'A', 'color': '#aaaaaa', 'hydrophobic': True, 'heatStable': True, 'toxinResistant': False},
    'Tyr': {'name': 'Tyrosine', 'short': 'Y', 'color': '#00e5ff', 'hydrophobic': True, 'heatStable': False, 'toxinResistant': True},
    'His': {'name': 'Histidine', 'short': 'H', 'color': '#00e5ff', 'hydrophobic': False, 'heatStable': False, 'toxinResistant': True},
    'Gln': {'name': 'Glutamine', 'short': 'Q', 'color': '#00ff88', 'hydrophobic': False, 'heatStable': False, 'toxinResistant': True},
    'Asn': {'name': 'Asparagine', 'short': 'N', 'color': '#00ff88', 'hydrophobic': False, 'heatStable': False, 'toxinResistant': False},
    'Lys': {'name': 'Lysine', 'short': 'K', 'color': '#ff3366', 'hydrophobic': False, 'heatStable': False, 'toxinResistant': False},
    'Asp': {'name': 'Aspartic Acid', 'short': 'D', 'color': '#ff3366', 'hydrophobic': False, 'heatStable': True, 'toxinResistant': False},
    'Glu': {'name': 'Glutamic Acid', 'short': 'E', 'color': '#ff3366', 'hydrophobic': False, 'heatStable': True, 'toxinResistant': True},
    'Cys': {'name': 'Cysteine', 'short': 'C', 'color': '#ffaa00', 'hydrophobic': True, 'heatStable': False, 'toxinResistant': True},
    'Trp': {'name': 'Tryptophan', 'short': 'W', 'color': '#a855f7', 'hydrophobic': True, 'heatStable': False, 'toxinResistant': False},
    'Arg': {'name': 'Arginine', 'short': 'R', 'color': '#ff3366', 'hydrophobic': False, 'heatStable': False, 'toxinResistant': True},
    'Gly': {'name': 'Glycine', 'short': 'G', 'color': '#aaaaaa', 'hydrophobic': False, 'heatStable': True, 'toxinResistant': False},
    'Stop': {'name': 'Stop', 'short': '*', 'color': '#666666', 'hydrophobic': False, 'heatStable': False, 'toxinResistant': False},
}

# Puzzle requirements dictionary matching targetTraits in Next.js puzzles.ts
PUZZLES = {
    'tv-01': {
        'target_traits': {'heat_resistance': 40},
        'min_stability': 40,
    },
    'tv-02': {
        'target_traits': {'heat_resistance': 55},
        'min_stability': 40,
    },
    'tv-03': {
        'target_traits': {'heat_resistance': 50, 'membrane_stability': 60},
        'min_stability': 40,
    },
    'tv-04': {
        'target_traits': {'heat_resistance': 70, 'membrane_stability': 55},
        'min_stability': 45,
    },
    'tv-05': {
        'target_traits': {'heat_resistance': 75, 'membrane_stability': 60, 'metabolic_efficiency': 50},
        'min_stability': 50,
    },
    'tm-01': {
        'target_traits': {'toxin_resistance': 40},
        'min_stability': 40,
    },
    'tm-02': {
        'target_traits': {'toxin_resistance': 55, 'membrane_stability': 45},
        'min_stability': 40,
    },
    'tm-03': {
        'target_traits': {'toxin_resistance': 65},
        'min_stability': 40,
    },
    'tm-04': {
        'target_traits': {'toxin_resistance': 70, 'membrane_stability': 55, 'metabolic_efficiency': 45},
        'min_stability': 45,
    },
    'tm-05': {
        'target_traits': {'toxin_resistance': 75, 'membrane_stability': 60, 'metabolic_efficiency': 55},
        'min_stability': 50,
    }
}

def translate_sequence(dna: str) -> list:
    """Translate DNA sequence into amino acids, stopping at a Stop codon."""
    amino_acids = []
    clean = "".join([c for c in dna.upper() if c in "ATCG"])
    
    for i in range(0, len(clean) - 2, 3):
        codon = clean[i:i+3]
        aa = CODON_TABLE.get(codon, '???')
        if aa == 'Stop':
            break
        amino_acids.append(aa)
        
    return amino_acids

def calculate_traits(amino_acids: list) -> dict:
    """Calculate trait percentages from amino acid properties."""
    heat_score = 0
    toxin_score = 0
    membrane_score = 0
    metabolic_score = 0
    
    for aa in amino_acids:
        props = AA_PROPERTIES.get(aa)
        if not props:
            continue
            
        if props['heatStable']:
            heat_score += 12
        if props['toxinResistant']:
            toxin_score += 12
        if props['hydrophobic']:
            membrane_score += 8
        else:
            membrane_score += 4
        metabolic_score += 6
        
    total = len(amino_acids) if amino_acids else 1
    
    return {
        'heat_resistance': min(100, round((heat_score / total) * 8)),
        'toxin_resistance': min(100, round((toxin_score / total) * 8)),
        'membrane_stability': min(100, round((membrane_score / total) * 10)),
        'metabolic_efficiency': min(100, round((metabolic_score / total) * 12)),
    }

def calculate_folding_stability(amino_acids: list) -> int:
    """Calculate protein folding stability from amino acid sequence."""
    if not amino_acids:
        return 0
        
    stability = 50
    hydrophobic_runs = 0
    current_run = 0
    
    for aa in amino_acids:
        props = AA_PROPERTIES.get(aa)
        if not props:
            continue
            
        if props['hydrophobic']:
            current_run += 1
            if current_run >= 3:
                hydrophobic_runs += 1
        else:
            current_run = 0
            
        if aa == 'Pro':
            stability += 3
        if aa == 'Cys':
            stability += 5
        if aa == 'Gly':
            stability -= 2
            
    stability += hydrophobic_runs * 5
    return max(0, min(100, stability))

def evaluate_organism(puzzle_id: str, sequence: str) -> dict:
    """Evalues if sequence matches target requirements and returns detailed diagnostics."""
    amino_acids = translate_sequence(sequence)
    traits = calculate_traits(amino_acids)
    stability = calculate_folding_stability(amino_acids)
    
    puzzle = PUZZLES.get(puzzle_id)
    if not puzzle:
        return {
            "survived": False,
            "failurePoint": "UNKNOWN PUZZLE",
            "failureReason": f"Puzzle '{puzzle_id}' does not exist.",
            "traits": traits,
            "stability": stability,
            "aminoAcids": amino_acids
        }
        
    # Check folding stability first
    min_stability = puzzle['min_stability']
    if stability < min_stability:
        return {
            "survived": False,
            "failurePoint": "PROTEIN FOLDING",
            "failureReason": f"Protein folding stability is {stability}%, but minimum required is {min_stability}%. Try introducing hydrophobic runs or Proline/Cysteine to stabilize the protein structure.",
            "failurePhase": "protein_folding",
            "traits": traits,
            "stability": stability,
            "aminoAcids": amino_acids
        }
        
    # Check trait requirements
    for trait_name, req_level in puzzle['target_traits'].items():
        actual_level = traits.get(trait_name, 0)
        if actual_level < req_level:
            # Generate contextual details
            label = trait_name.replace("_", " ").title()
            return {
                "survived": False,
                "failurePoint": label,
                "failureReason": f"Insufficient {label}. Calculated level is {actual_level}%, but environmental threshold requires {req_level}%. Modify nucleotide codons to express stronger trait-specific properties.",
                "failurePhase": trait_name,
                "traits": traits,
                "stability": stability,
                "aminoAcids": amino_acids
            }
            
    # All checks passed
    return {
        "survived": True,
        "traits": traits,
        "stability": stability,
        "aminoAcids": amino_acids
    }
